### ISSUE TITLE
`global` is undefined in SSR rendering

### ISSUE THREAD

**Issue by davestewart** [phoenixframework/phoenix#4833]
### Intro

Possibly something you won't want to fix, but – tracking down some badly written code in a project I've just joined – Phoenix is being run during server side generation (SSG) and failing with a JS error due to the unexpected environment.

During instantiation [the code references](https://github.com/phoenixframework/phoenix/blob/master/priv/static/phoenix.js#L766) the `global` object...:

```js
this.transport = opts.transport || global.WebSocket || LongPoll;
```

...which [is undefined](https://github.com/phoenixframework/phoenix/blob/master/priv/static/phoenix.js#L35) in this case:

```js
 var global = globalSelf || phxWindow || void 0;
```

I feel it would be better to fail gracefully here, or defend against it.

I guess it really depends on what your intention is to support the user, or where you feel the user should have used their knowledge to not be in this situation.

### Environment

* Elixir version (elixir -v): client only
* Phoenix version (mix deps): 1.6.8
* Operating system: OSX

**Comment by chrismccord** [comment]
In your case, what is defined as "global" in your environment? We look for `globalSelf`, `window`, or `global`. You can also pass the transport directly with `{transport: MyWhateverScope.WebSocket}`

**Comment by chrismccord** [comment]
We should fix the global reference, but failing to find websocket in node is expected. Node has no global/native `WebSocket` client, so you need to require/import the dep and pass it to the socket constructor instead: 

```javascript
import ws from "whatever-latest/websocket-client"
let socket = new Socket("/socket", {transport: ws})
```

**Comment by davestewart** [comment]
Hi Chris,

Great! Exactly what I was thinking.

I'll get that implemented. Thanks for your help!

**Comment by chrismccord** [comment]
It's not clear to me how you ended up with the compiled line because the tagged released contains this:

https://github.com/phoenixframework/phoenix/blob/master/priv/static/phoenix.cjs.js#L45 (notę: no void: 0). Did you build the js on your side? We can change the global shadowing, but I don't think there's actually anything to fix on our side. If you can provide more info, let me know. Thanks!

**Comment by davestewart** [comment]
Hi Chris,

I'm looking at the code now as well, and you're right, it's not there.

Pretty damn sure I pasted the code verbatim, just a normal NPM install for us.

We've actually removed SSR from our project now, so won't be able to test it (though I could check out an old branch of course) but as mentioned, it should just be an installed package, not something clever we've done.

The only thing I can think is that maybe WebStorm's debugger replaced `undefined` with `void 0` ?

Anyway, I reread the post and the main point was about the `global` shadowing anyway, so I'll leave you to consider that.

Cheers,
Dave
